pkg/controllers/metrics/termination/controller.go

```diff
@@ -0,0 +1,139 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package termination
+
+import (
+	"context"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/prometheus/client_golang/prometheus"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/metrics"
+	"github.com/aws/karpenter/pkg/utils/functional"
+)
+
+const (
+	controllerName = "terminationmetrics"
+	provisioner    = "provisioner"
+)
+
+var (
+	terminationProvisionerSummaryVec = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace:  "karpenter",
+			Subsystem:  "nodes",
+			Name:       "termination_provisioner_time_seconds",
+			Help:       "The time taken between a node's deletion request and its termination, labeled by provisioner",
+			Objectives: metrics.SummaryObjectives(),
+		},
+		[]string{provisioner},
+	)
+
+	terminationSummaryVec = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace:  "karpenter",
+			Subsystem:  "nodes",
+			Name:       "termination_time_seconds",
+			Help:       "The time taken between a node's deletion request and its termination",
+			Objectives: metrics.SummaryObjectives(),
+		},
+		[]string{},
+	)
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(terminationProvisionerSummaryVec)
+	crmetrics.Registry.MustRegister(terminationSummaryVec)
+}
+
+type Controller struct {
```

Comment:
I worry a bit about the additional processing and memory of this controller. We're caching the full cluster's nodes here, and reconciling on every single node object update. Does it make more sense to put these metrics in the termination controller, where we're directly handling the termination of nodes?